### PR TITLE
Fix test trade

### PR DESCRIPTION
**Pull request by wakamex**
merge into dev this time

### CODE REVIEW

File: src/elfpy/markets.py [hunk 1]
```diff
@@ -175,3 +177,18 @@ def get_market_state_string(self):
...
+                share_reserves=self.share_reserves,
+                bond_reserves=self.bond_reserves,
+                init_share_price=self.init_share_price,
+                share_price=self.share_price,
+                time_remaining=time_utils.stretch_time(
+                    self.token_duration, self.time_stretch_constant
+                )
+            )
+        else:
+            self.spot_price = np.nan
```
Comment (dpaiton) [inline src/elfpy/markets.py:194]: add newline at end of file (this will resolve linting issues later)

File: src/elfpy/simulators.py [hunk 2]
```diff
@@ -295,6 +298,7 @@ def run_simulation(self, override_dict=None):
                 f"\ninit_time_stretch = {self.market.time_stretch_constant}"
                 "\n-----\n"
             )
+        last_user_action_time = 0
```
Comment (dpaiton) [inline src/elfpy/simulators.py:301]: Can't you pull this variable directly from the analysis dict?

File: src/elfpy/simulators.py [hunk 3]
```diff
@@ -408,4 +445,4 @@ def update_analysis_dict(self):
             self.analysis_dict["fee"].append(self.fee * self.base_asset_price)
             slippage = (self.without_fee_or_slippage - self.without_fee) * self.base_asset_price
             self.analysis_dict["slippage"].append(slippage)
-        # self.analysis_dict["spot_price"].append(self.market.spot_price())
+        self.analysis_dict["spot_price"].append(self.market.spot_price)
```
Comment (dpaiton) [inline src/elfpy/simulators.py:448]: add newline

File: src/elfpy/simulators.py [hunk 4]
```diff
@@ -322,23 +326,56 @@ def run_simulation(self, override_dict=None):
...
+                            #     f"\n\tuser trade = {user_action}"
+                            #     f"\n\tuser_wallet = {user.wallet}"
+                            #     f"\n\tinit pool apy = {self.init_pool_apy}"
+                            #     f"\n\ttrades = {self.market.base_asset_orders + self.market.token_asset_orders}"
+                            #     f"\n\tinit_share_price = {self.market.init_share_price}"
+                            #     f"\n\tshare_price = {self.market.share_price}"
+                            #     f"\n\tamount = {self.trade_amount}"
+                            #     f"\n\treserves = {(self.market.share_reserves, self.market.bond_reserves)}"
+                            # )
+                if (self.market.time - last_user_action_time > 0.1) and (self.market.time - last_user_action_time) % 0.5 < 1/365/self.num_blocks_per_day/2:
```
Comment (dpaiton) [inline src/elfpy/simulators.py:354]: Shouldn't this be in a verbose conditional? If so, can you also include a TODO above that reminds us to fix this up when we improve logging.

File: src/elfpy/strategies/basic.py [hunk 5]
```diff
@@ -14,6 +14,10 @@ class BasicPolicy(User):
     most basic policy setup
     """
 
+    def __init__(self, market, rng, budget=1000, verbose=False):
```
Comment (dpaiton) [inline src/elfpy/strategies/basic.py:17]: Can you fix the arg order so it matches the parent class

File: src/elfpy/strategies/single_short.py [hunk 6]
```diff
@@ -1,36 +1,46 @@
...
+        self.mint_times = list(self.wallet["token_in_wallet"].keys())
+        self.has_opened_short = True if any([x < -1 for x in self.position_list]) else False
+        self.can_open_short = self.get_max_short(self.market.time) >= self.amount_to_trade
+
+    def status_report(self):
+        return (
+            f"has_opened_short: {self.has_opened_short}, can_open_short: {self.can_open_short}"
+            + f" max_short: {self.get_max_short(self.market.time)}"
+            + f" position_list: {self.position_list} sum(positions)={sum(self.position_list)}"
+        )
```
Comment (dpaiton) [inline src/elfpy/strategies/single_short.py:46]: add newline

File: tests/notebook_test.ipynb [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
```
Comment (dpaiton) [inline tests/notebook_test.ipynb:42]: clear outputs & commit again

File: tests/test_trade.py [hunk 8]
```diff
@@ -56,13 +56,19 @@ def run_base_trade_test(self, policy):
...
 
-    def test_base_user(self):
+    def test_single_long(self):
         """Tests the BaseUser class"""
         self.run_base_trade_test("single_long")
-        self.run_base_trade_test("single_short")
+
+    def test_single_short(self):
+        """Tests the BaseUser class"""
+        self.run_base_trade_test("single_short")
```
Comment (dpaiton) [inline tests/test_trade.py:74]: add newline

File: tests/test_trade.py [hunk 9]
```diff
@@ -56,13 +56,19 @@ def run_base_trade_test(self, policy):
             "fee_percent": 0.1,
             "init_pool_apy": 0.05,
         }
+        if additional_overrides:
```
Comment (dpaiton) [inline tests/test_trade.py:59]: Can you specify it in the same way as override_dict, and then use override_dict.update(additional_overrides)?

File: tests/test_trade.py [hunk 10]
```diff
@@ -15,7 +15,7 @@
 class BaseTradeTest(unittest.TestCase):
     """Generic Trade Test class"""
 
-    def run_base_trade_test(self, policy):
+    def run_base_trade_test(self, policy, additional_overrides=None):
```
Comment (dpaiton) [inline tests/test_trade.py:18]: where is this actually used?

File: src/elfpy/simulators.py (hunk 2 above)
Comment (wakamex) [reply src/elfpy/simulators.py:301]: ya, but not really needed to store. it's not part of the simulation output, i used it only to do logging when i hadn't seen a trade for a while. it's not used in the final configuration (trades happen often enough). so this should be removed and rolled into proper logging. but it's useful to keep for basic debugging for now.

File: src/elfpy/strategies/basic.py (hunk 5 above)
Comment (wakamex) [reply src/elfpy/strategies/basic.py:17]: moved verbose to the end

File: tests/test_trade.py (hunk 10 above)
Comment (wakamex) [reply tests/test_trade.py:18]: in notebook_test, to allow quick testing in the notebook only, without having to edit the config file, which also runs the risk of having to revert it when you're done, or forgetting what you've changed

File: tests/test_trade.py (hunk 9 above)
Comment (wakamex) [reply tests/test_trade.py:59]: cleaned it up, commented below where it's used

File: src/elfpy/strategies/basic.py (hunk 5 above)
Comment (dpaiton) [inline src/elfpy/strategies/basic.py:17]: This function is no longer needed if it is only a passthrough.

File: src/elfpy/user.py [hunk 11]
```diff
@@ -117,3 +130,19 @@ def update_wallet(self, trade_result):
                     pass
                 else:
                     raise ValueError(f"key={key} is not allowed.")
+        wallet_string = ""
```
Comment (dpaiton) [inline src/elfpy/user.py:133]: Is this a verbose output? Can you add a todo to make it such?

**Pull request merged**